tb_pulumi/__init__.py

```diff
@@ -208,3 +214,37 @@ def env_var_is_true(name: str) -> bool:
     """
 
     return env_var_matches(name, ['t', 'true', 'yes'], False)
+
+
+def flatten(item: dict | list | ThunderbirdComponentResource | pulumi.Resource) -> set[pulumi.Resource]:
+    """Recursively traverses a nested collection of Pulumi ``Resource``s, converting them into a flat set which can be
+    more easily iterated over.
+
+    :param item: Either a Pulumi ``Resource`` object, or some collection thereof. The following types of collections are
+        supported: ``dict``, ``list``, ``ThunderbirdComponentResource``.
+    :type item: dict | list | ThunderbirdComponentResource
+
+    :return: A ``set`` of Pulumi ``Resource``s contained within the collection.
+    :rtype: set(pulumi.Resource)
+    """
+
+    # The item could be of a variety of types. When the item is some kind of collection, we should compress it down into
+    # a flat list first, then operate on its items.
+    flattened = []
+    to_flatten = None
+    if type(item) is list:
+        to_flatten = item
+    elif type(item) is dict:
+        to_flatten = [value for _, value in item.items()]
+    elif isinstance(item, ThunderbirdComponentResource):
+        to_flatten = [value for _, value in item.resources.items()]
+    elif isinstance(item, pulumi.Resource):
+        return [item]
+    else:
+        pass
```

Comment:
You can run into `None`s and a bunch of other stuff we don't care about. It's easiest to just ignore them all with this dumb "else/pass".